### ISSUE TITLE
SAMFire documentation 

### ISSUE THREAD

**Issue by mwalls** [hyperspy/hyperspy#1628]
As requested by Tomas I will list a few things that might be worth mentioning in the docs. The text below, or parts of it, could be a starting point:
1) _Note! When importing data that have been binned using other software, the stored values may be the averages of the binned channels or pixels, instead of their sum, as would be required for proper statistical analysis. While Hyperspy will try to reconstruct the correct values using the metadata _(not yet implemented...)_, we cannot guarantee that the statistics will be valid. We therefore strongly recommend that all pre-fitting binning should be done using Hyperspy_
2) In the part where you talk about tolerance, mention that if you change it you should refresh the database and that you can put it to np.inf to see if the programme will run at all.
3) After: "_As a rule of thumb, the first strategy in the list should always be from the local family, followed by a strategy from the global family_." I'd put: _"Indeed, for the time being, except for the most experienced users, we recommend using only the local strategy. This can be implemented via samf.remove(1)"_

I'll stop there for the moment. Not sure whether the policy is to have one item per issue or to group related ones. Also not sure if you'd prefer to write the actual version yourself or leave it to users like me who are using it on their own data...
I could try and do it but I'm not great at inserting links in the docs as yet. (Can't be bothered to get over the entry barrier of Sphinx for the moment).

**Comment by tjof2** [comment]
I don't think point 1 is exclusively a Samfire issue.

**Comment by mwalls** [comment]
You're right , So I've made a pull request which puts it in the general curve-fitting part of the docs

**Comment by thomasaarholt** [comment]
I would love for someone to upload a SAMFire jupyter notebook.

**Comment by to266** [comment]
@thomasaarholt by the time the notebook is "finished" to be shared, it would look kind of trivial (just writing from memory, so things may be misspelled):
```python
s = hs.load("......mydata.something")
s.estimate_poissonian_noise_variance()
s.metadata.Signal.Noise_parameters.variance.data += 1 #rarely needed with raw data
# s.set_microscope_parameters(200, 25, 30)
m = s.create_model()
# set up your model here....
# ...
# ...
m.axes_manager.indices = (15,3) # go to a good starting index
fitting_args  = dict(bounded=True)
m.fit(**fitting_args)
samf = m.create_samfire(workers=number)
samf.metadata.goodness_test.tolerance = 7.3
samf.remove(1) # likely you don't want histogram
samf.refresh_database()
# samf.plot()
samf.start(**fitting_args)
#...
m.plot_results()
m.save('/path/to/my_model_filename.hdf5', 'samfire_model_one')
```

The main difficulty, I found, is setting everything up - sometimes the variance does not physically make sense (is 0) and the weights get infinite, sometimes the tolerance has to be increased, sometimes you need a better / different starting pixel, and sometimes to bound the parameters more or less.

Thus closer to an ideal solution would be a video of someone (I guess me) doing all this and recording, hopefully with a voice-over and explaining, and also sharing the data and the finished notebook.

The main difficulty is the dataset:
 - Should be open-access, and preferably not hosted by us (something more reliable).
 - The analysis should actually __require__ SAMFire - i.e. it should be complex / difficult enough that traditional `multifit` fails and/or diverges.
 - Hopefully the analysis is not too specific and the actual model is close to obvious (a couple of Gaussians for two peaks). The goal would be to help as many people as possible, with as familiar a problem as we can manage

**Comment by thomasaarholt** [comment]
Good points.

Suggestions for the[ SAMFire documentation](http://hyperspy.org/hyperspy-doc/current/user_guide/model.html#smart-adaptive-multi-dimensional-fitting-samfire).
- Clarify under "Seed Points" that a seed pixel is created by calling m.fit() in that pixel, and that the user should move around on the plot to each new region and call m.fit() there. *Before* creating samfire (or calling refresh_database). I am still not sure if what I just said is correct, so it certainly needs clarification.

- You don't explain what the two strategies are - I can imagine what the red_chisq is, but no idea what the histogram does.

- What does the "database" mean in "The current strategy “database” can be plotted using the plot() method."?

- Clarify why the red_chisq value needs to be between 0 and 2? 

- What does the tolerance value mean/do? What is its max and min? "Use a sensible value" isn't really clear :)

- Update "mpfit" in the final line to be "leastsq".

**Comment by mwalls** [comment]
I agree with nearly all of this. I too do not know what the database is or how the strategies work really. For the Chisq, being a fairly standard concept, maybe a link to some site might be worthwhile for those who are not familiar with it, rather than an explanation of the HS site?

**Comment by magnunor** [comment]
Even having a very, very simple example would be great. As currently I'm not able to get it working, and I'm not sure if it is due to a bug, or because I'm doing something wrong.

It could be as simple as making an [artificial dataset](http://hyperspy.org/hyperspy-doc/current/api/hyperspy.datasets.artificial_data.html#module-hyperspy.datasets.artificial_data), with two different "structural phases". One "phase" could consist of a Gaussian at x=20 and the other of a Gaussian at x=40. Just so people can test the functionality, and see it working.

**Comment by ericpre** [comment]
@magnunor: I guess once you managed to get it to work, you should be in a good position to provide the example? :)

**Comment by thomasaarholt** [comment]
@ericpre I think @magnunor is saying he _hasn't_ been able to make it work :) I've given it a few tries over the years as well, and have not been able either.

**Comment by magnunor** [comment]
@thomasaarholt, @ericpre, indeed, I tried for a little while, but couldn't get it running. My comment was kind of a response to @to266 comment above:

> The analysis should actually require SAMFire - i.e. it should be complex / difficult enough that traditional multifit fails and/or diverges.

Where I think having an extremely simple examples is fine, since people can easily take that simple example and tweak it for their own data.

SAMFire is one of the features in HyperSpy I'd really, really like to use. But I don't work with that type of 2-D navigation dims. and 1-D signal dims. often enough nowadays to dedicate the time to figure out how to get it working.

**Comment by tjof2** [comment]
@thomasaarholt  and @magnunor  when you say you've not made it work, can you be more specific? 

Related SAMFire issues #1277, #1347, #2110.

Is it really the case that the tests in https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/hyperspy/tests/samfire/test_samfire.py don't cover the basics?

@dnjohnstone do you have any input on this?
